src/presentation/controllers/device-controller.ts

```diff
@@ -0,0 +1,131 @@
+              const incfg = config as INewsDevice
+              return new INewsDeviceRestDTO(config, incfg.username, incfg.password)
+            case 'TelemetricsDevice': // Corrected case label
+              // eslint-disable-next-line no-case-declarations
+              const tmcfg = config as TelemetricsDevice
+              return new TelemetricsDeviceRestDTO(config, tmcfg.host)
+            default:
+              throw new Error(`Unknown device type: ${config.type}`)
+          }
+        })
+      ))
+    } catch (error) {
+      this.httpErrorHandler.handleError(response, error as Exception)
+    }
+  }
+
+
+  @GetRequest('/:id')
+  public async getDeviceConfiguration(request: Request, response: Response): Promise<void> {
+    try {
+      const deviceId = request.params.id
+      const config = await this.deviceService.readConfiguration(deviceId)
+      response.send(this.httpResponseFormatter.formatSuccessResponse(new DeviceDto(config)))
+    } catch (error) {
+      this.httpErrorHandler.handleError(response, error as Exception)
+    }
+  }
+
+  @PostRequest('/')
+  public async createDeviceConfiguration(request: Request, response: Response): Promise<void> {
+    try {
+      const deviceDto: DeviceRestDto = request.body as DeviceRestDto //discriminator field
+
+      switch(deviceDto.type){
+        case 'INewsDevice':
+          await this.deviceService.create(request.body as INewsDevice)
+          break
+        case 'TelemetricsDevice':
+          await this.deviceService.create(request.body as TelemetricsDevice)
+          break
+        default:
+          console.log('createDeviceConfiguration rececived a malformed request') //Help, replace me with a proper logger
+          break
+      }
+  
+      response.send(this.httpResponseFormatter.formatSuccessResponse())
+    } catch (error) {
+      this.httpErrorHandler.handleError(response, error as Exception)
+    }
+  }
+
+  @PutRequest('/:id')
+  public async updateDeviceConfiguration(request: Request, response: Response): Promise<void> {
+    try {
+      const deviceId = request.params.id
+      const config = request.body
+      await this.deviceService.update(deviceId, config)
+      response.status(200).json(config)
+    } catch (error) {
+      this.httpErrorHandler.handleError(response, error as Exception)
+    }
+  }
+
+  @DeleteRequest('/:id')
+  public async deleteDeviceConfiguration(request: Request, response: Response): Promise<void> {
+    try {
+      const deviceId = request.params.id
+      await this.deviceService.delete(deviceId)
+      response.status(204).send()
+    } catch (error) {
+      this.httpErrorHandler.handleError(response, error as Exception)
+    }
+  }
+
+  @PostRequest('/:id/reconnect')
+  public async manualReconnectDevice(request: Request, response: Response): Promise<void> {
+    try {
+      const deviceId = request.params.id
+      await this.deviceService.disconnect(deviceId)
+      await this.deviceService.connect(deviceId)
```

Comment:
It does happen through the service layer, but I'm guessing you want the composition of these events done by a wrapper method in the logic layer that is called once, instead of manually orchestrating them from the controller, amirite?

Reply:
Yes. It might be the service layer that does the actually disconnecting and connectiong, but in this implementation, it is the presentation layer that knows that a reconnect requiers a disconnect followed by a connect, which is logic. The presentation layer should just validate its input and then call a method in another layer to handle the logic for it.

Reply:
Code not present in codebase any longer.